### ISSUE TITLE
encoding/protobuf: generated Go code does not compile

### ISSUE THREAD

**Issue by cueckoo** [cue-lang/cue#581]
_Originally opened by @dwilson1988 in https://github.com/cuelang/cue/issues/581_

### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version devel linux/amd64
</pre>

### Does this issue reproduce with the latest release?
Yes

### What did you do?
Created a proto file for my dataset. This can be reproduces with the following proto:

"dataset.proto"
```protobuf
syntax = "proto3";
package protos;

import "cue/cue.proto";
option go_package = "my.org/protos";

message Test {
  string test = 1;
}
```

I ran protoc on this with the following commands:
```bash
include = ${GOPATH}/src/cuelang.org/go/encoding/protobuf

protoc \
  -I=./protos \
  -I=${include} \
  --go_out=./protos \
  --go_opt=paths=source_relative \
  dataset.proto
```
Note: I tried withouth source_relative as well.

This created the following (abridged) dataset.pb.go:
```golang
// Code generated by protoc-gen-go. DO NOT EDIT.
// versions:
// 	protoc-gen-go v1.25.0
// 	protoc        v3.6.1
// source: dataset.proto

package protos

import (
	_ "cuelang.org/cueproto"
	proto "github.com/golang/protobuf/proto"
	protoreflect "google.golang.org/protobuf/reflect/protoreflect"
	protoimpl "google.golang.org/protobuf/runtime/protoimpl"
	reflect "reflect"
	sync "sync"
)

const (
	// Verify that this generated code is sufficiently up-to-date.
	_ = protoimpl.EnforceVersion(20 - protoimpl.MinVersion)
	// Verify that runtime/protoimpl is sufficiently up-to-date.
	_ = protoimpl.EnforceVersion(protoimpl.MaxVersion - 20)
)

// This is a compile-time assertion that a sufficiently up-to-date version
// of the legacy proto package is being used.
const _ = proto.ProtoPackageIsVersion4

type Test struct {
	state         protoimpl.MessageState
	sizeCache     protoimpl.SizeCache
	unknownFields protoimpl.UnknownFields

	Test string `protobuf:"bytes,1,opt,name=test,proto3" json:"test,omitempty"`
}
```

### What did you expect to see?
The problem is the import  "cuelang.org/cueproto" refers to a package that's not go-getable.
```
$ go get cuelang.org/cueproto
go get cuelang.org/cueproto: unrecognized import path "cuelang.org/cueproto": reading https://cuelang.org/cueproto?go-get=1: 404 Not Found
```

### What did you see instead?
I would expect the generated pb to refer to a path that I could access, but that module doesn't appear to be accessible. 

I'm guessing that refers to a package under "cuelang.org/go/encoding/protobuf", as that's where the proto file was, but that refers to it's go_package as "cuelang.org/cueproto".  

Perhaps I am missing something, but I am stumped.

**Comment by cueckoo** [comment]
_Original reply by @verdverm in https://github.com/cuelang/cue/issues/581#issuecomment-721322816_

Where are you doing your protobuf work from?

Are you accidentally picking up this file? https://github.com/cuelang/cue/blob/master/encoding/protobuf/cue/cue.proto

**Comment by cueckoo** [comment]
_Original reply by @dwilson1988 in https://github.com/cuelang/cue/issues/581#issuecomment-724341312_

@verdverm, sorry, somehow I didn't notice this...

I didn't make this clear in my original description, but I'm applying cue constraints in my actually application and using that to generate a schema. E.g.:

```protobuf
message Dataset {
  // machine parsable dataset name
  string name = 1 [(cue.val) = '=~"^[a-z0-9]+(?:-[a-z0-9]+)*$"'];
}
```

So I guess I'm not accidentally picking up that file, that's an explicit import since I need it for my constraints/schema gen.

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/581#issuecomment-731576367_

This is probably related to #450.

The respective proto is in `encoding/protobuf/cue/cue.proto`, so you'll have to include that in your path when evaluating.

Suggestions on how to make this clearer or what would be a better approach is welcome.

**Comment by myitcv** [comment]
@dwilson1988 thanks for raising this, and for the steps to reproduce. I've consolidated this into a standalone reproducer (details on how to create repros like this [in the wiki](https://github.com/cue-lang/cue/wiki/Creating-test-or-performance-reproducers)):

```
# Set explicit GOMODCACHE
env GOMODCACHE=$WORK/modcache

# Create a go.sum
go mod tidy

# Explicitly download the cuelang.org/go dependency
go mod download cuelang.org/go

# Run protoc.
#
# Ideally the -I flag to protoc would be derived from a command
# output like:
#
#     go list -f {{.Dir}} cuelang.org/go/encoding/protobuf
#
# but testscript does not support sub-shells.
exec protoc -I $GOMODCACHE/cuelang.org/go@v0.4.0/encoding/protobuf -I . --go_out=. --go_opt=module=my.org/protos basic.proto

# Verify generated Go output exists and is tidy
exists basic.pb.go
go mod tidy

# Ensure the code compiles
go test

-- basic.proto --
syntax = "proto3";

// Package basic is just that: basic.
package protos;

import "cue/cue.proto";

option go_package = "my.org/protos";

// This is my type.
message MyType {
  string string_value = 1; // just any 'ole string

  // A method must start with a capital letter.
  repeated string method = 2 [ (cue.val) = '[...=~"^[A-Z]"]' ];
}
-- deps.go --
// +build protodeps

package protodeps

import (
	_ "cuelang.org/go/encoding/protobuf"
)
-- go.mod --
module my.org/protos

go 1.16

require cuelang.org/go v0.4.0
```

The output from this is as follows:

```
# Set explicit GOMODCACHE (0.000s)
> env GOMODCACHE=$WORK/modcache
# Create a go.sum (4.074s)
> go mod tidy
[stderr]
go: downloading cuelang.org/go v0.4.0
...

# Explicitly download the cuelang.org/go dependency (0.146s)
> go mod download cuelang.org/go
# Run protoc.
#
# Ideally the -I flag to protoc would be derived from a command
# output like:
#
#     go list -f {{.Dir}} cuelang.org/go/encoding/protobuf
#
# but testscript does not support sub-shells. (0.032s)
> exec protoc -I $GOMODCACHE/cuelang.org/go@v0.4.0/encoding/protobuf -I . --go_out=. --go_opt=module=my.org/protos basic.proto
# Verify generated Go output exists and is tidy (2.257s)
> exists basic.pb.go
> go mod tidy
[stderr]
...
go: finding module for package cuelang.org/cueproto
my.org/protos imports
        cuelang.org/cueproto: cannot find module providing package cuelang.org/cueproto: unrecognized import path "cuelang.org/cueproto": reading https://cuelang.org/cueproto?go-get=1: 404 Not Found
```

This is slightly different from the original output, but the reason is that this repro is using Go 1.16 which no longer automatically tidies a `go.mod`. So despite the fact the `go test` step never runs, we can be sure it would not compile/exit successfully because the `go mod tidy` step fails to resolve an import.

My analysis so far:

`basic.pb.go` is generated by `protoc` (plus `protoc-gen-go`), taking as input `basic.proto` and the import `cue/cue.proto`. `basic.pb.go` includes the following import for side effects:

```
import _ "cuelang.org/cueproto"
```

This is driven by the fact that `cue/cue.proto` defines its Go package as that path:

https://github.com/cue-lang/cue/blob/0f53054da7243c7a312cd0ad54a1083d67c185d3/encoding/protobuf/cue/cue.proto#L7

This guidance on the use of `cue/cue.proto` comes from the [protobuf integration documentation](https://cuelang.org/docs/integrations/protobuf/#extract-cue-from-a-standalone-proto-file), which demonstrates how the imported `cue` package can be used as a field option to specify a more precise constraint in CUE that might be generated from the proto source.

However, unlike the protobuf integration example, this example generates Go code from the proto source (those two are clearly not mutually exclusive). So the missing "piece" here is, as @mpvl points out, exactly what https://github.com/cuelang/cue/pull/450 originally sought to do: ensure that Go code generated from proto that has CUE-oriented field options actually compiles. Plus tests to lock in the behaviour of course.

My sense therefore is that https://github.com/cuelang/cue/pull/450 was probably heading in the right direction:

* We should distribute the equivalent of the currently generated `cuelang.org/cueproto` as part of the CUE distribution.
* It doesn't make sense (for now at least) to version that separately, so it can live as part of the `cuelang.org/go` module.
* Per @mpvl's comment in that PR, it probably makes sense to make the Go package generated from `cue/cue.proto` have the path `cuelang.org/go/encoding/protobuf/cueproto`. Which raises the question, perhaps the proto source itself should live at `encoding/protobuf/cueproto/cue.proto`? Moving `cue/cue.proto` would be a breaking change, so necessary to split across a couple of releases with good release notes.

To accompany this we need better documentation on how to ensure that `protoc` can/should be used in conjunction with the `cuelang.org/go` dependency. Because fundamentally, anyone needing to `import "cue/cue.proto"` (as it exists today) needs a dependency on `cuelang.org/go` in order to resolve that file via a `-I` directive, so shipping the Go generated result as part of that same Go module really comes at no cost, only benefit. At least I'm not aware of a better way to resolve the `cue/cue.proto` dependency today. 

The one question that arises from all of this is what version of `protoc` and `protoc-gen-go` to use to generate the result of `cue/cue.proto` (wherever it lives). The best answer I can come up with there is "the latest or as close as possible to it". Clearly we would pin the version in CI for reproducibility's sake.

TL;DR - generating Go code from a proto file annotated with `cue` field options does not work today, and we need to fix that.

**Comment by karlmutch** [comment]
For others that might wander here. I am using cue annotations in protobuf to be able to extract the cue rather than to use cue encodings in protobuf.

The very limited workaround I am using is to copy in the cue.proto file into my own repo and use it as a stub in my own source tree.  This works if your not interested in anything other than having a single file contain the cue and proto for your application but does not address the root use-case of course.

**Comment by randomizedcoder** [comment]
Looks like there's been no movement on this issue in a long time, but I do see "generate source code from CUE definitions (TODO)" in the readme, so that's good.
